### DISCUSSION TITLE
Switching snippets, assets and templates together with the theme?

### DISCUSSION THREAD

**Discussion by richi** [datenstrom/community#206] [Ask a question] [answered]
In a lot of cases one has to modify the snippets etc. to create more sophisticated themes. Unfortunately this does often break other themes. Would it be possible to implement another structure like the following:

themes
..default
....snippets
....templates
....assets
....default.css
..theme1
....snippets
....templates
....assets
....theme1.css
..theme2
....snippets
....templates
....assets
....theme2.css

First the engine would look for a file in the selected theme folder, if it is not there the one from the default folder comes to use.

I guess this would make life a lot easier for theme developers. What do you think?

Cheers, richi

**Comment by BsNoSi** [comment]
I love Yellow for its simplicity. Personally, I never needed the option, to switch between layouts. Therefore, each theme in a ZIP overwriting existing one while unpacking, would be enough for me. This does not require any adjust to config aso., without required additional code for offering a in my opinion very seldom requirement.

**Comment by markseuffert** [comment]
I would like to understand what you want to do, Richi. Can you explain a bit more please. How does your work flow look like as a theme developer? Which problems do you experience? Do you use multiple themes on one website? Do you make themes for different websites and need a test bed?

**Comment by intermedion** [comment]
I'd very much like the suggested structure in a new version. This would allow to easily develop and test new themes or to modify existing ones, without breaking another theme. The current structure makes this unnecessarily hard and offers no advantages over a well organized structure, like the one suggested.

**Comment by markseuffert** [comment]
I very much like the idea of tools adjusting to your way of working, instead of you having to adjust to a tool. Can you explain a bit more how a new structure would make your development easier? This is not meant to criticize, but to learn more what developers like to do.

**Comment by nibreh** [comment]
That's thrue, in a classic CMS, generally there is just to move the theme folder containing new HTML structure and CSS styles, and apply it. Each theme are separated and goes with layout and CSS, because both are connected.

Yellow does not allow to use different themes with a different HTML layout. For those who like to develop themes, it's a bit strange.

With Richi's suggestion, maybe you got more files because of each theme folders, but you have the power to develop each theme separately.

I hope I have been clear and not mistaken, my English is bad :)

**Comment by BsNoSi** [comment]
I do understand the attempt of switching between several templates. But I have in mind, what the main use of Yellow shall be: a neat, robust and stable website „generator“.

Therefore I take it as allowed question, if a process, typically only relevant during creation process should rise several additional challenges on operation, i.e. handling of backups, management of updates, a.s.o.

In my opinion, there is no  no elaborated change required. If you want to switch between different templates, you can do it currently quite simple:

Change the structure into

- themes
    - theme
        - templates
        - snippets
        - assets
        - additional files
    - theme1
        - templates
        - snippets
        - assets
        - additional files
    - …
 
 Then you can simply switch between themes by modifying the path in config.ini to `themes/theme1` … . Easiest way is creating the required amount of ini-files, config.ini1, config.ini2 (a.s.o.), that can be switched really fast by renaming (i.e. command line batch).

**Comment by richi** [comment]
A lot of people use Yellow as a full blown CMS. As an example: To get http://www.emuseum-tettnang.de online we had to change much more than the CSS. Having done that, one can't switch (easily ) to a foreign theme (on sub pages) anymore and updates are somewhat fragile as well. Not to mention allowing the user to switch the theme... 

Having everything belonging to a theme in one folder and switching this folder instead of just the css would not add complexity to Yellow, in contrary I guess it would make life easier for most developers.

Of course, BsNoSi is right, one can get around this using the entires in config.ini, but that looks a little hackish and there is still no "default" folder.

Having a default folder with the "official" templates, snippets etc. would actually reduce complexity, because a theme folder would only need to include the changed files. Of course distribution of themes would be easier and more transparent (IMO).

**Comment by BsNoSi** [comment]
I didn´t mention, that you can change `theme1`…`themex` to `default`. this is extremely close to your entry at top of this thread (I noticed that after posting). 

I think the question is, if the „standard install“ should offer a „multi theme enviroment“. Maybe there is a `Theme : themex` entry in file header a solution, that points to a subfolder of the designed structure of Rici (that I closely copied a few entries later).

**Comment by markseuffert** [comment]
Thanks Richi for your explanation. I think Yellow should offer this kind of flexibility and I have a simple solution in mind. I can't promise anything, but will have a look when the next release has come out.

If someone would like to see this feature, please add +1 to the top of this issue.

**Comment by markseuffert** [comment, marked as answer]
Sorry for keeping you waiting.

With the help of the Yellow API you can now switch assets/snippets/templates together with the theme.  This is an experimental feature which you could try out. Just keep in mind that this might be changed or removed at anytime. Tested with Yellow 0.6.8.

Create a new file `system/plugins/multitheme.php`:

```
<?php

class YellowMultitheme
{
	const VERSION = "0.6.1";
	var $yellow;			//access to API

	// Handle initialisation
	function onLoad($yellow)
	{
		$this->yellow = $yellow;
		$path = $this->yellow->config->get("themeDir");
		foreach($this->yellow->toolbox->getDirectoryEntries($path) as $entry)
		{
			$this->yellow->themes->load($entry."/assets/");
		}
	}
	
	// Handle page meta data parsing
	function onParseMeta($page)
	{
		if($page==$this->yellow->page)
		{
			$themeDir = $this->yellow->config->get("themeDir").$page->get("theme")."/";
			$this->yellow->config->set("assetDir", $themeDir."assets/");
			$this->yellow->config->set("snippetDir", $themeDir."snippets/");
			$this->yellow->config->set("templateDir", $themeDir."templates/");
			$themeLocation = $this->yellow->config->get("themeLocation").$page->get("theme")."/";
			$this->yellow->config->set("assetLocation", $themeLocation."assets/");
		}
	}
}

$yellow->plugins->register("multitheme", "YellowMultitheme", YellowMultitheme::VERSION);
?>
```
Go to your `system/themes` folder and create a new folder for each theme. Here's an example:

![screen shot 2017-01-05 at 22 15 48](https://cloud.githubusercontent.com/assets/2429165/23094587/2528eccc-f5fc-11e6-901b-6321167b1b04.png)

Hope this is what you are looking for.

**Comment by richi** [comment]
I love this solution! That's just perfect...
